### PR TITLE
feat(kad): migrate to quick-protobuf-codec crate

### LINKED ISSUE #4487: kad: migrate to `quick-protobuf-codec` crate
In `/misc` we have the `quick-protobuf-codec` crate which makes working with protobuf structs over streams quite easy as you only need to construct a `asynchronous_codec::Framed` type and can read and write full protobuf messages on any `Async{Read,Write}`.

We should make use of this in `libp2p-kad`.

See https://docs.rs/quick-protobuf-codec.

Previous tracking issue: https://github.com/libp2p/rust-libp2p/issues/2500

### PR DESCRIPTION
**Pull request by 0xcrust**
## Description

Resolves #4487.

## Notes & open questions
Much thanks to @thomaseizinger for the very useful pointers!

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/kad/CHANGELOG.md [hunk 1]
```diff
@@ -4,9 +4,11 @@
   See [PR 4270].
 - Reduce noise of "remote supports our protocol" log.
   See [PR 4278].
+- Depend on `quick-protobuf-codec` crate for codec logic.
```
Comment (thomaseizinger) [inline protocols/kad/CHANGELOG.md]: Version `0.44.4` has already been released. Please make a new entry above! See https://github.com/libp2p/rust-libp2p/blob/master/CONTRIBUTING.md#write-changelog-entries-for-user-facing-changes and the linked release docs! :)

File: protocols/kad/src/protocol.rs [hunk 2]
```diff
@@ -179,6 +177,47 @@ impl UpgradeInfo for KademliaProtocolConfig {
     }
 }
 
+/// Codec for Kademlia inbound and outbound message framing.
+pub struct KadCodec<A, B> {
```
Comment (thomaseizinger) [inline protocols/kad/src/protocol.rs]:
```suggestion
pub struct Codec<A, B> {
```

We are within the scope of `libp2p-kad` so I'd say this is redundant. Related: https://github.com/libp2p/rust-libp2p/issues/4485.

File: protocols/kad/src/protocol.rs [hunk 3]
```diff
@@ -179,6 +177,47 @@ impl UpgradeInfo for KademliaProtocolConfig {
...
+}
+
+impl<A: Into<proto::Message>, B> Encoder for KadCodec<A, B> {
+    type Error = io::Error;
+    type Item = A;
+
+    fn encode(&mut self, item: Self::Item, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        self.codec
+            .encode(item.into(), dst)
+            .map_err(|err| err.into())
```
Comment (thomaseizinger) [inline protocols/kad/src/protocol.rs]: I know that `?` automatically calls `From::from` for you so I think you might be able to get away with using `?` here instead of `map_err`! :)

File: protocols/kad/src/protocol.rs [hunk 4]
```diff
@@ -179,6 +177,47 @@ impl UpgradeInfo for KademliaProtocolConfig {
...
+    }
+}
+impl<A, B: TryFrom<proto::Message, Error = io::Error>> Decoder for KadCodec<A, B> {
+    type Error = io::Error;
+    type Item = B;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        self.codec
+            .decode(src)?
+            .map(|msg| B::try_from(msg))
```
Comment (thomaseizinger) [inline protocols/kad/src/protocol.rs]: I think clippy might want you to write the following:

```suggestion
            .map(B::try_from)
```

File: protocols/kad/src/protocol.rs [hunk 5]
```diff
@@ -179,6 +177,47 @@ impl UpgradeInfo for KademliaProtocolConfig {
...
+            .transpose()
+    }
+}
+
+/// Sink of responses and stream of requests.
+pub(crate) type KadInStreamSink<S> = Framed<S, KadCodec<KadResponseMsg, KadRequestMsg>>;
+/// Sink of requests and stream of responses.
+pub(crate) type KadOutStreamSink<S> = Framed<S, KadCodec<KadRequestMsg, KadResponseMsg>>;
+
```
Comment (thomaseizinger) [inline protocols/kad/src/protocol.rs]: Are these still useful now? Or should we just delete them?

File: protocols/kad/src/protocol.rs [hunk 6]
```diff
@@ -188,32 +227,9 @@ where
     type Error = io::Error;
```
Comment (thomaseizinger) [inline protocols/kad/src/protocol.rs:222]: Just seeing that we actually never return this error! Meaning we can probably simplify a few things here by replacing this with `void::Void`. Should be done in a separate PR though to keep the scope small. Let me know if you are interested :)

File: protocols/kad/src/protocol.rs [hunk 7]
```diff
@@ -346,6 +323,31 @@ pub enum KadResponseMsg {
...
+        proto_to_req_msg(message)
+    }
+}
+impl TryFrom<proto::Message> for KadResponseMsg {
+    type Error = io::Error;
+
+    fn try_from(message: proto::Message) -> Result<Self, Self::Error> {
+        proto_to_resp_msg(message)
+    }
+}
```
Comment (thomaseizinger) [inline protocols/kad/src/protocol.rs:344]: Is it worth having this code? As far as I can see, you are still calling all of them manually in the codec, right? I don't think we gain anything from having these then and I'd rather not have these 20 LoC and instead call the functions they are delegating too directly from the codec.

File: protocols/kad/CHANGELOG.md (hunk 1 above)
Comment (thomaseizinger) [reply protocols/kad/CHANGELOG.md]: You will also need to bump the version in the `Cargo.toml`!

**Review by thomaseizinger — COMMENTED** [review]
Thank you!

Looks good overall, I've left some minor comments :)

File: protocols/kad/CHANGELOG.md (hunk 1 above)
Comment (thomaseizinger) [reply protocols/kad/CHANGELOG.md]: > You will also need to bump the version in the `Cargo.toml`!

And also the referenced version in the root `Cargo.toml` because we use workspace inheritance to sync the versions of our crates across the workspace.

File: protocols/kad/src/protocol.rs (hunk 5 above)
Comment (0xcrust) [reply protocols/kad/src/protocol.rs]: Not necessarily, but it does prevent us having to write the more lengthy `Framed<Stream, Codec<KadResponseMsg, KadRequestMsg>>` multiple times in `handler.rs`

File: protocols/kad/CHANGELOG.md (hunk 1 above)
Comment (0xcrust) [reply protocols/kad/CHANGELOG.md]: Done! Would `0.44.5` be the next logical bump?

File: protocols/kad/src/protocol.rs (hunk 2 above)
Comment (0xcrust) [reply protocols/kad/src/protocol.rs]: got it. fixed now

File: protocols/kad/src/protocol.rs (hunk 6 above)
Comment (0xcrust) [reply protocols/kad/src/protocol.rs:222]: I'd love to send in a PR for that immediately after this!

File: protocols/kad/src/protocol.rs (hunk 7 above)
Comment (0xcrust) [reply protocols/kad/src/protocol.rs:344]: That would mean we have to define both `InCodec` and `OutCodec` structs, with different `Encoder` & `Decode` trait impls since we can't rely on just switching the positions of `KadResponseMsg` & `KadRequestMsg` on the same type.

File: protocols/kad/CHANGELOG.md (hunk 1 above)
Comment (thomaseizinger) [reply protocols/kad/CHANGELOG.md]: > Done! Would `0.44.5` be the next logical bump?

Yes!

File: protocols/kad/src/protocol.rs (hunk 5 above)
Comment (thomaseizinger) [reply protocols/kad/src/protocol.rs]: Okay, happy to keep them then!

File: protocols/kad/src/protocol.rs (hunk 7 above)
Comment (thomaseizinger) [reply protocols/kad/src/protocol.rs:344]: Ah you are generic over these! Sorry didn't see that, that is actually quite clean. Nicely done :)

File: protocols/kad/src/protocol.rs (hunk 4 above)
Comment (0xcrust) [reply protocols/kad/src/protocol.rs]: Thanks! I just pushed a commit applying this and the other suggestions above

File: protocols/kad/CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,9 @@
+## 0.44.5
```
Comment (thomaseizinger) [inline protocols/kad/CHANGELOG.md]:
```suggestion
## 0.44.5 - unreleased
```

File: protocols/kad/CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,9 @@
...
+- Migrate to `quick-protobuf-codec` crate for codec logic.
+  see [PR 4501].
```
Comment (thomaseizinger) [inline protocols/kad/CHANGELOG.md]:
```suggestion
  See [PR 4501].
```

File: protocols/kad/CHANGELOG.md [hunk 10]
```diff
@@ -7,6 +13,7 @@
 
 [PR 4270]: https://github.com/libp2p/rust-libp2p/pull/4270
 [PR 4278]: https://github.com/libp2p/rust-libp2p/pull/4278
+[PR 4501]: https://github.com/libp2p/rust-libp2p/pull/4501
```
Comment (thomaseizinger) [inline protocols/kad/CHANGELOG.md]:
```suggestion
```

File: protocols/kad/Cargo.toml [hunk 11]
```diff
@@ -21,6 +21,7 @@ log = "0.4"
 libp2p-core = { workspace = true }
 libp2p-swarm = { workspace = true }
 quick-protobuf = "0.8"
```
Comment (thomaseizinger) [inline protocols/kad/Cargo.toml:23]: Can we remove this one now?

**Review by thomaseizinger — APPROVED** [review]
Looks great! Thank you :)

**Comment by thomaseizinger** [comment]
Seems you missed to checkin changes to the lockfile :)

**Review by mxinden — COMMENTED** [review]
Very clean! Thank you for the work.

File: protocols/kad/Cargo.toml (hunk 11 above)
Comment (0xcrust) [reply protocols/kad/Cargo.toml:23]: It's still used in `src/generated/dht/pb.rs` so probably not

**Review by thomaseizinger — DISMISSED** [review]

File: protocols/kad/src/protocol.rs [hunk 12]
```diff
@@ -28,19 +28,17 @@
...
 use instant::Instant;
 use libp2p_core::upgrade::{InboundUpgrade, OutboundUpgrade, UpgradeInfo};
 use libp2p_core::Multiaddr;
 use libp2p_identity::PeerId;
 use libp2p_swarm::StreamProtocol;
-use quick_protobuf::{BytesReader, Writer};
+use std::marker::PhantomData;
 use std::{convert::TryFrom, time::Duration};
 use std::{io, iter};
-use unsigned_varint::codec;
```
Comment (thomaseizinger) [inline protocols/kad/src/protocol.rs:43]: Only just noticed that I think we can remove this dependency now, right? Mind sending a follow-up?

Comment (0xcrust) [reply protocols/kad/src/protocol.rs:43]: `unsigned_varint`? It's already removed

edit: ah you mean from Cargo.toml. Sure I can do that in a follow-up

**Review by thomaseizinger — APPROVED** [review]
